### PR TITLE
[RFC] vim-patch:7.4.609

### PR DESCRIPTION
**Pull request by oni-link**
vim-patch:7.4.609

Problem:    For complicated list and dict use the garbage collector can run
                 out of stack space.
Solution:   Use a stack of dicts and lists to be marked, thus making it
            iterative instead of recursive. (Ben Fritz)

https://github.com/vim/vim/commit/2459a5ecaa43c8549ea53e9364253ff891676da5

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -5711,61 +5712,143 @@ static int free_unref_items(int copyID)
...
-      set_ref_in_ht(&dd->dv_hashtab, copyID);
+    case VAR_DICT: {
+      dict_T *dd = tv->vval.v_dict;
+      if (dd != NULL && dd->dv_copyID != copyID) {
+        // Didn't see this dict yet.
+        dd->dv_copyID = copyID;
+        if (ht_stack == NULL) {
+          abort = set_ref_in_ht(&dd->dv_hashtab, copyID, list_stack);
+        } else {
+          ht_stack_T *newitem = try_malloc(sizeof(ht_stack_T));
```
Comment (oni-link) [inline src/nvim/eval.c]: Would `xmalloc()` be enough here, or do we expect to run out of heap memory here?

**Comment by oni-link** [comment]
Test script to be used with `nvim -S test.vim -u NONE`:
```vim
if !exists('g:iter')
  let g:iter=1000000
endif

fun! CyclicDictionary(count)
  let start = {}
  let prev = start
  for i in range(1, a:count)
    let new_ = {'prev': prev}
  endfor
  let start.prev = new_

  return start
endf

fun! CyclicList(count)
  let lists = map(range(0,a:count-1), '[]')
  for i in range(0, a:count-1)
    call add(lists[i], lists[(i+1) % a:count])
    call add(lists[i], lists[(i-1+a:count) % a:count])
  endfor
  return lists[0]
endf

let start = reltime()
let dict = CyclicDictionary(g:iter)
echomsg "Dict ready: ".string(reltimestr(reltime(start)))
let start = reltime()
let list = CyclicList(g:iter)
echomsg "List ready: ".string(reltimestr(reltime(start)))
let start = reltime()
call garbagecollect(1)
echomsg "Garbage collect: ".string(reltimestr(reltime(start)))
```

File: src/nvim/eval.c (hunk 1 above)
Comment (justinmk) [inline src/nvim/eval.c]: `garbage_collect()` may be called just before exit, so I suggest leaving this as it is.

**Comment by justinmk** [comment]
Test results:

Nvim (without this patch), OS X 10.10.5:

```
Dict ready: '  3.286615'
List ready: ' 16.819202'
Garbage collect: '  0.000015'
<segfault>
```

Vim 7.3.0, , OS X 10.10.5:

```
Dict ready: '  3.552861'
List ready: ' 10.010602'
Garbage collect: '  0.000023'
<segfault>
```

Nvim (with this patch), OS X 10.10.5:

```
Dict ready: '  3.489032'
List ready: ' 16.116184'
Garbage collect: '  0.000026'
<no segfault>
```

Similar results on 32-bit ubuntu VM.

LGTM

**Pull request merged**